### PR TITLE
Von Mises Fisher Loss and decoding

### PR DESCRIPTION
**Pull request by artidoro**

### CODE REVIEW

File: evaluation.py [hunk 1]
```diff
@@ -52,7 +53,11 @@ def eval(model, loss_function, test_iter, args):
 
     for batch in tqdm.tqdm(test_iter):
         scores = model.forward(batch.src, batch.trg)
-        loss = loss_function(scores[:-1,:,:].view(-1, scores.shape[2]), batch.trg[1:,:].view(-1))
+        if isinstance(loss_function, VonMisesFisherLoss):
```
Comment (artidoro) [inline evaluation.py]: Same comment as in train.py

File: train.py [hunk 2]
```diff
@@ -16,7 +17,11 @@ def train(model, optimizer, scheduler, loss_function, train_iter, val_iter, args
         for batch in tqdm.tqdm(train_iter):
             optimizer.zero_grad()
             scores = model(batch.src, batch.trg)
-            loss = loss_function(scores[:-1,:,:].view(-1, scores.shape[2]), batch.trg[1:,:].view(-1))
+            if isinstance(loss_function, VonMisesFisherLoss):
```
Comment (brendon-boldt) [inline train.py:20]: @artidoro I had forgotten earlier why I had this conditional here. It is because the loss function needs to be passed different values with cross-entropy vs vMF loss. This is probably the neatest way to do things since the loss function won't have access to the embeddings otherwise.

**Comment by brendon-boldt** [comment]
Closes #4

**Comment by brendon-boldt** [comment]
Also, the conda enivonrment specification which I included in this works. I was able to use it to set up a success environment on "my" server (although you still have to manually run spaCy commands).

File: evaluation.py [hunk 3]
```diff
@@ -49,13 +55,29 @@ def eval(model, loss_function, test_iter, args):
...
+            # Remove first elem of batch.trg which is start token.
+            target = model.decoder.embedding(batch.trg[1:,:].view(-1))
+        else:
+            target = batch.trg[1:,:].view(-1)
+        loss = loss_function(scores[:-1,:,:].view(-1, scores.shape[2]), target)
         loss_tot += loss.item()
-        preds = scores[:-1,:,:].argmax(2).squeeze()
-        correct += sum((preds.view(-1) == batch.trg[1:,:].view(-1))).item()
+        if is_vmf_loss:
+            # TODO bb: I dont' know why the last elem is removed.
```
Comment (artidoro) [inline evaluation.py]: We don't care about the token that is predicted after the end of sentence.
```
t1    t2   t3  <\s>  (garbage)
 |     |     |     |     |
<s>   t1    t2     t3    <\s>
```

File: evaluation.py [hunk 4]
```diff
@@ -73,39 +95,56 @@ def eval(model, loss_function, test_iter, args):
...
     """
-    predictions, ground_truth = greedy_decoding(model, test_iter, args)
-    bleu = {'blue': sacrebleu.corpus_bleu(predictions, [ground_truth])}
-    if args['write_to_file']:
-        write_predictions(predictions, args)
-    return bleu
\ No newline at end of file
+    predictions, ground_truth = greedy_decoding(model, test_iter, max_decoding_len)
+    bleu = {"bleu": sacrebleu.corpus_bleu(predictions, [ground_truth])}
+    print(bleu["bleu"].score)
```
Comment (artidoro) [inline evaluation.py]: Can you remove the print statements?
Add to the dictionary if you want the results to be logged (or write_predictions if you want them in a file).

File: evaluation.py [hunk 5]
```diff
@@ -73,39 +95,56 @@ def eval(model, loss_function, test_iter, args):
...
-                logits = model(batch.src, decoded_sentences)
-                decoded_sentences[current_time_step,:,:] = vocab_id[logits.argmax(-1)]
-                current_time_step += 1
-                still_decoding = torch.all(torch.any(torch.argmax(decoded_sentences, dim = -1) == EOS_IDX, dim = 0))
-            predictions += list(logits.argmax(-1).transpose(1,0).tolist())
+            ground_truth += batch.trg.transpose(1, 0).tolist()
+            predictions += model.decode(batch.src, max_decoding_len)
+
+    # TODO What is this?
+    # model.train(mode)
```
Comment (artidoro) [inline evaluation.py]: Can you put this back?
It will revert the eval/train flag of the model to the original mode (so if the model was in training mode it will revert to that). I think it helps to know that the functions will not alter the state of the model.

File: main.py [hunk 6]
```diff
@@ -113,9 +117,28 @@ def parse_args(args):
...
+            model,
+            optimizer,
+            scheduler,
+            loss_function,
+            train_iter,
+            val_iter,
+            args
+        )
+        evaluation.decode(model, val_iter, args['max_decoding_len'])
+        # def decode(model, test_iter, args):
```
Comment (artidoro) [inline main.py]: Can you remove the comment?

File: model.py [hunk 7]
```diff
@@ -134,51 +147,98 @@ def forward(self, query, values):
...
+        self.global_attn = GlobalAttention(
+            query_size=kwargs["dec_hidden_size"],
+            values_size=kwargs["enc_hidden_size"] * (2 ** kwargs["enc_bidirectional"]),
+            out_size=kwargs["dec_embed_size"],
+        )
         if self.xent:
-            self.linear1 = nn.Linear(kwargs['dec_embed_size'], vocab_size)
-        self.dropout = nn.Dropout(kwargs['dropout'])
+            self.linear1 = nn.Linear(kwargs["dec_embed_size"], vocab_size)
+        # TODO Stop dropout at eval time.
```
Comment (artidoro) [inline model.py]: That's taken care already. When you do model.eval() if you use nn.Dropout() it will stop dropping out.
That's why I asked you put that back in the decoding function.

File: model.py [hunk 8]
```diff
@@ -134,51 +147,98 @@ def forward(self, query, values):
...
+        batch_size = h_encoder.shape[1]
+        decoded_embeds = [bos_embed.reshape(1, -1).repeat(1, batch_size, 1)]
+        model_out = output = torch.zeros(
+            [1, batch_size, self.embedding.weight.shape[1]]
+        ).to(self.embedding.weight.device)
+        hidden = None
+        decoded_idxs = [np.array([[bos_idx]]).repeat(batch_size, 1)]
+        eos_generated = np.zeros((1, batch_size), dtype=np.bool)
+        while len(decoded_idxs) < max_decoding_len and (eos_generated == 0).any():
+            model_out, hidden = self.step(  ###
```
Comment (artidoro) [inline model.py]: What are the three ###?

File: model.py [hunk 9]
```diff
@@ -134,51 +147,98 @@ def forward(self, query, values):
...
+                decoded_embeds[-1], model_out, hidden, h_encoder,
+            )
+            if self.xent:
+                model_sm = self.linear1(model_out)
+                decoded_idxs.append(model_sm.argmax(-1).cpu().numpy())
+                decoded_embeds.append(self.embedding(model_sm.argmax(-1)))
```
Comment (artidoro) [inline model.py]: Instead of calling the embedding function here and above for bos token, can you do it at the beginning of the while loop before calling step?

File: model.py [hunk 10]
```diff
@@ -134,51 +147,98 @@ def forward(self, query, values):
...
+        bos_idx_tensor = torch.LongTensor([bos_idx]).to(self.embedding.weight.device)
+        bos_embed = self.embedding(bos_idx_tensor)
+        batch_size = h_encoder.shape[1]
+        decoded_embeds = [bos_embed.reshape(1, -1).repeat(1, batch_size, 1)]
+        model_out = output = torch.zeros(
+            [1, batch_size, self.embedding.weight.shape[1]]
+        ).to(self.embedding.weight.device)
+        hidden = None
+        decoded_idxs = [np.array([[bos_idx]]).repeat(batch_size, 1)]
+        eos_generated = np.zeros((1, batch_size), dtype=np.bool)
```
Comment (artidoro) [inline model.py:220]: That's a cool trick, very smart!

**Review by artidoro — COMMENTED** [review]
Just a few minor comments. But looks great!

**Comment by artidoro** [comment]
Funy thing is that I cannot approve this PR because I was the one filing it hahaha good to know for the future.

**Comment by brendon-boldt** [comment]
@artidoro Thanks for the suggestions. If it looks good, just comment so (if you can't merge your own PR).

File: evaluation.py [hunk 11]
```diff
@@ -73,39 +94,47 @@ def eval(model, loss_function, test_iter, args):
...
+
+def decode(
+    model, test_iter, max_decoding_len, write_to_file=False, checkpoint_path=None,
+) -> Dict:
+    # TODO: Think about returning more than just bleu (ex: ppx, loss, ...).
+    predictions, ground_truth = greedy_decoding(model, test_iter, max_decoding_len)
+    bleu = {"bleu": sacrebleu.corpus_bleu(predictions, [ground_truth])}
+    logger = logging.getLogger("vivo_logger")
+    logger.info(f"Test BLEU Score: {bleu['bleu'].score:.2f}")
+    if write_to_file:
```
Comment (artidoro) [inline evaluation.py:135]: I think we log the eval results in main.

File: model.py [hunk 12]
```diff
@@ -134,51 +147,105 @@ def forward(self, query, values):
...
             else:
                 outputs.append(output)
         return torch.cat(outputs, 0)
 
+    def decode(
+        self, h_encoder: Tensor, max_decoding_len: int, bos_idx: int, eos_idx: int,
+    ) -> List[List[int]]:
+        bos_idx_tensor = torch.LongTensor([bos_idx]).to(self.embedding.weight.device)
+        batch_size = h_encoder.shape[1]
+        decoded_embeds = []
```
Comment (artidoro) [inline model.py]: Why do you keep the list?

**Review by artidoro — COMMENTED** [review]
Just another two minor things but you can check it in after.
